hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/JoinNestedLoopPhysicalRel.java

```diff
@@ -109,6 +114,36 @@ public <V> V accept(CreateDagVisitor<V> visitor) {
         return visitor.onNestedLoopJoin(this);
     }
 
+    /**
+     * Cost calculation of Nested Loop Join relation.
+     * NLJ algorithm is simple algorithm, where for each left row we are traversing right row set.
+     * Speaking of cost estimation, we are accounting the next properties:
+     * - row count is estimating ans L * R, because for each left row we're probing full right row set;
+     * - same for CPU cost estimation multiplied by cost of row comparison.
+     * The perfect assumption also must include memory and IO cost estimation, and also
+     * a selectivity for a right row set.
+     */
+    @Override
+    @Nullable
+    public RelOptCost computeSelfCost(RelOptPlanner planner, RelMetadataQuery mq) {
+        final double leftRowCount = mq.getRowCount(left);
+        final double rightRowCount = mq.getRowCount(right);
+        if (Double.isInfinite(leftRowCount) || Double.isInfinite(rightRowCount)) {
+            return planner.getCostFactory().makeInfiniteCost();
+        }
+
+        RelOptCost rightCost = planner.getCost(getRight(), mq);
+        if (rightCost == null) {
+            return planner.getCostFactory().makeInfiniteCost();
+        }
+
+        // TODO: introduce selectivity estimator, but ATM we taking the worst case scenario : selectivity = 1.0.
+        double rowsEstimate = leftRowCount * /* TODO: selectivity */ rightRowCount;
+        double cpuEstimate = Math.max(1.0, rowsEstimate - 1) * Cost.JOIN_ROW_CMP_MULTIPLIER;
```

Comment:
you did not include cost of repetition of right query, which is repeated leftRowCount times (possibly -1 because first repetition is included in default calculation on the rel tree)

Reply:
Considering that these are rough estimates, will it make a noticeable difference to use `Math.max(1.0, rowsEstimate - 1)` instead of just `rowsEstimate`?

Reply:
> you did not include cost of repetition of right query, which is repeated leftRowCount times 

row higher:  `double rowsEstimate = leftRowCount *  rightRowCount;`

Reply:
>  will it make a noticeable difference to use Math.max(1.0, rowsEstimate - 1) instead of just rowsEstimate?

Yes, if you have one row, you'll get zero cost, what is unlikely.